File: templates/problems/13_dp_1d/05_longest_palindrome_sub/typescript/solution.ts
/**
 * Return the longest palindromic substring.
 * Time: O(n^2), Space: O(1)
 */
export function solveLongestPalindrome(s: string): string {
    if (s.length < 2) return s;
    let start = 0;
    let maxLen = 1;

    const expand = (l: number, r: number): void => {
        while (l >= 0 && r < s.length && s[l] === s[r]) {
            if (r - l + 1 > maxLen) {
                start = l;
                maxLen = r - l + 1;
            }
            l--;
            r++;
        }
    };

    for (let i = 0; i < s.length; i++) {
        expand(i, i); // odd length
        expand(i, i + 1); // even length
    }
    return s.slice(start, start + maxLen);
}

File: templates/problems/13_dp_1d/05_longest_palindrome_sub/python/solution.py
def solve_longest_palindrome(s: str) -> str:
    """Return the longest palindromic substring.

    Time: O(n^2), Space: O(1)
    """
    if len(s) < 2:
        return s
    start, max_len = 0, 1

    def expand(l: int, r: int) -> None:
        nonlocal start, max_len
        while l >= 0 and r < len(s) and s[l] == s[r]:
            if r - l + 1 > max_len:
                start = l
                max_len = r - l + 1
            l -= 1
            r += 1

    for i in range(len(s)):
        expand(i, i)  # odd length
        expand(i, i + 1)  # even length
    return s[start : start + max_len]
